### ISSUE TITLE
[Review] Checkers_vibe — C++ REST API + Node.js + React

### ISSUE THREAD

**Issue by ArgonZapan** [ArgonZapan/Checkers_vibe#4]
## Code Review — Checkers_vibe

**Autor:** Jarvis (AI QA)  
**Commity:** 8 commits — C++ REST API, Node.js server, React frontend, integracja WebSocket

---

## Ocena ogólna: 🟢 Solidna architektura — dobry flow danych

---

### Architektura
- Czyste 3-warstwy: C++ (silnik) → Node.js (AI, proxy, socket.io) → React (frontend)
- socket.io do real-time komunikacji — `state`, `move`, `gameOver` events
- TensorFlow.js do neural network (modelWhite/modelBlack per side)
- Replay buffer (10000 samples) z auto-save co 10 min

### Node.js server
- Express routes: `/api/ai/*` (TensorFlow), `/api/selfplay/*` (game loop)
- Proxy do C++ dla `/api/game/*`, `/api/legal-moves`, `/api/move`
- WebSocket: `startGame`, `move`, `getLegalMoves`, `aiMove` helper
- Auto-save modeli co 5 min

### React frontend
- `EMPTY_BOARD()` — poprawna inicjalizacja 8x8, ciemne pola na (r+c)%2==1
- `handleCellClick` — selekcja pionka + wysyłanie `getLegalMoves`
- `handleMove` — czyste oddzielenie: cell click → emit move
- Tryby: menu → PvAI → AIvsAI + Dashboard

### Model AI (model.js)
- Input: 257 features (64×4 channels + turn indicator)
- 4 kanały: empty, white, black, king
- Policy head: 48 neuronów (max ruchów), softmax
- Value head: tanh (-1 do 1)
- Epsilon-greedy z dekayem

---

### 1. Błędna asignacja `result` w `_playGame` (trainer.js)
```js
s.result = s.turn === (result > 0 ? 1 : result < 0 ? -1 : 0) ? Math.abs(result) : -Math.abs(result);
```
**Problem:** `result` to string `'white'|'black'|'draw'`, nie liczba. Operator `>` na stringu daje `false`, więc `result > 0` → `false`, `result < 0` → `false`. Wynik: `result > 0 ? 1 : ...` → zawsze `1`. Czyli:
- Wszystkie próbki dostają `result = ±1` tak samo, bez względu na to czyj ruch wygrał
- Black samples nigdy nie dostają ujemnego result

**Poprawka:**
```js
const winnerTurn = result === 'white' ? 1 : result === 'black' ? -1 : 0;
for (const s of samples) {
  s.result = s.turn === winnerTurn ? 1 : -1;
}
```

### 2. `lastMove` pokazuje ruch gracza zamiast AI (server/index.js, linia ~move handler)
```js
socket.emit('state', {
  ...state,  // state ma lastMove od C++ (poprzedni ruch, nie AI)
  lastMove: { from, to },  // to jest ruch GRACZA, nie AI!
});
```
**Problem:** Gracz widzi swój własny ruch jako "ostatni", zamiast zobaczyć ruch AI.

**Poprawka:**
```js
socket.emit('state', {
  ...state,
  lastMove: { from: selectedMove.from, to: selectedMove.to },  // ruch AI
});
```

---

### 3. Buffer `shift()` jest O(n) — memory leak potential
```js
add(sample) {
  this.buffer.push(sample);
  while (this.buffer.length > this.maxSize) {
    this.buffer.shift();  // O(n) — przesuwa wszystkie elementy
  }
}
```
**Problem:** `Array.shift()` kopiuje całą tablicę. Przy 10000 elementów i częstym add, to może być wolne.

**Sugestia:** Użyj indeksu cyklicznego:
```js
this.buffer = new Array(maxSize);
this.head = 0;
add(sample) {
  this.buffer[this.head] = sample;
  this.head = (this.head + 1) % this.maxSize;
  this.buffer.length = Math.min(this.buffer.length + 1, this.maxSize);
}
```

### 4. Brak timeout na fetch do C++ (trainer.js `_playGame`)
```js
const startRes = await fetch(`${CPP_BASE}/api/game/start`, { method: 'POST' });
```
**Problem:** Jeśli C++ backend nie odpowiada (np. crash), `fetch` będzie czekał bardzo długo (lub nigdy nie zwróci).

**Sugestia:** Dodać `AbortController` z timeoutem ~5s.

### 5. Proxy filter jest martwy kod (proxy.js)
```js
filter: (pathname) => {
  return !pathname.startsWith('/ai/') && !pathname.startsWith('/selfplay/');
}
```
**Problem:** W Express routes `/api/ai/*` i `/api/selfplay/*` są zdefiniowane PRZED `setupProxy()`. Więc te requesty nigdy nie docierają do proxy — filter nigdy nie jest sprawdzany.

### 6. `CPP_BASE` zdefiniowane ale nie używane (server/index.js)
```js
const CPP_BASE = 'http://localhost:8080';  // defined
// ... ale nie używane w tym pliku
```
Martwa stała — jest w `trainer.js` jako `const CPP_BASE = 'http://localhost:8080'` (poprawnie).

---

### 7. Brak walidacji `moveIndex` w `aiMove`
```js
const moveIndex = prediction.move;
let selectedMove = legalMoves[moveIndex] || legalMoves[0];
```
Jeśli `prediction.move` jest undefined lub poza zakresem, może wybrać zły ruch.

### 8. `gameMode` nie jest resetowany przy disconnect
Socket rozłącza się bez resetowania `gameMode`. Nowy klient może dziedziczyć stary tryb.

---

## Podsumowanie

| Aspekt | Ocena |
|--------|-------|
| Architektura | 🟢 Bardzo dobra |
| WebSocket flow | 🟢 Poprawne |
| TensorFlow integration | 🟢 Poprawne |
| AI self-play loop | 🔴 1 krytyczny bug (result assignment) |
| Memory/Performance | 🟠 buffer.shift() O(n) |
| Error handling | 🟠 Brak timeout na C++ fetch |

**Werdykt:** Architektura bardzo dobra. Muszą naprawić bug #1 (result assignment) bo sieć nigdy nie nauczy się poprawnie — black i white dostają takie same rewardy.

**Rekomendacja:** Naprawić #1 przed jakimkolwiek training. Reszta to optymalizacje.

---

*Recenzja wykonana przez Jarvis AI (QA).*

**Comment by ArgonZapan** [comment]
Naprawione:

1. ✅ Result assignment w trainer.js — string→number fix, draws=0, prawidłowe rewardy per strona
2. ✅ lastMove — teraz pokazuje ruch AI, nie gracza
3. ✅ buffer.shift() O(n) → splice O(excess)
4. ✅ cppFetch() z 5s timeout (AbortController)

Syntax checks OK.
